google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/startup/UnsupportedJavaVersionCheck.java

```diff
@@ -0,0 +1,177 @@
+import com.intellij.openapi.module.EffectiveLanguageLevelUtil;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.module.ModuleManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.LanguageLevelModuleExtension;
+import com.intellij.openapi.roots.ModifiableRootModel;
+import com.intellij.openapi.roots.ModuleRootManager;
+import com.intellij.openapi.startup.StartupActivity;
+import com.intellij.pom.java.LanguageLevel;
+import com.intellij.psi.xml.XmlFile;
+import com.intellij.psi.xml.XmlTag;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.swing.event.HyperlinkEvent;
+
+/**
+ * A StartupActivity that warns the user if they are using an unsupported java version.
+ */
+public class UnsupportedJavaVersionCheck implements StartupActivity {
+
+  private static final String RUNTIME_TAG = "runtime";
+  private static final String RUNTIME_TAG_JAVA_8 = "java8";
+  private static final String UPDATE_HREF = "#update";
+  private static final LanguageLevel HIGHEST_SUPPORTED_LANGUAGE_LEVEL = LanguageLevel.JDK_1_7;
+
+  @Override
+  public void runActivity(@NotNull Project project) {
+    List<Module> invalidModules = findAppEngineModulesUsingUnsupportedLanguageLevel(project);
+    if (!invalidModules.isEmpty()) {
+      warnUser(project, invalidModules);
+    }
+  }
+
+  private List<Module> findAppEngineModulesUsingUnsupportedLanguageLevel(Project project) {
+    Module[] projectModules = ModuleManager.getInstance(project).getModules();
+    List<Module> invalidModules = new ArrayList<>();
+
+    for (Module module : projectModules) {
+      // if it's not an app engine module, skip it
+      if (!hasAppEngineFacet(module)) {
+        continue;
+      }
+
+      @Nullable
+      XmlFile appengineWebXml = AppEngineAssetProvider.getInstance()
+          .loadAppEngineStandardWebXml(project, Arrays.asList(module));
+
+      if (isAppEngineStandard(appengineWebXml)
+          && usesJava8OrGreater(module)
+          && !declaresJava8Runtime(appengineWebXml)) {
+        invalidModules.add(module);
+      }
+    }
+    return invalidModules;
+  }
+
+  private boolean usesJava8OrGreater(Module module) {
+    LanguageLevel languageLevel = EffectiveLanguageLevelUtil.getEffectiveLanguageLevel(module);
+    return languageLevel.compareTo(LanguageLevel.JDK_1_8) >= 0;
+  }
+
+  private boolean hasAppEngineFacet(Module module) {
+    return AppEngineFacet.getAppEngineFacetByModule(module) != null;
+  }
+
+  private boolean isAppEngineStandard(@Nullable XmlFile appengineWebXml) {
+    AppEngineEnvironment environment = AppEngineProjectService.getInstance()
+        .getModuleAppEngineEnvironment(appengineWebXml);
+
+    return environment == AppEngineEnvironment.APP_ENGINE_STANDARD;
+  }
+
+  private boolean declaresJava8Runtime(@Nullable XmlFile appengineWebXml) {
```

Comment:
Sounds good. I was kind of on the fence about whether or not this was too much of a special case to belong in the DefaultAppEngineProjectService, but I'm happy to move it there.